File: tests/test_multiple_inheritance.py
import pytest
from common import collect
import test_multiple_inheritance_ext as t


def test_registry_smoke_miss_before_registration():
    assert t.registry_miss_before_registration()


def test_registry_smoke_register_and_hit():
    assert t.registry_register_and_lookup_hit()


def test_registry_smoke_unrelated_pair_remains_miss():
    assert t.registry_unrelated_pair_is_miss()


def test_registry_smoke_register_via_shim_helper():
    assert t.registry_register_with_shim_helper()


def test_smoke_shim_does_not_change_non_mi_behavior():
    obj = t.MIBase()
    assert obj.value() == 123
    assert t.shim_construct_and_call() == 123


def test_python_mi_isinstance_and_issubclass():
    d = t.MID()
    assert isinstance(d, t.MIB)
    assert isinstance(d, t.MIC)
    assert issubclass(t.MID, t.MIB)
    assert issubclass(t.MID, t.MIC)


def test_python_mi_mro_contains_all_bases():
    mro = t.MID.__mro__
    assert mro[0] is t.MID
    assert t.MIB in mro
    assert t.MIC in mro
    assert t.MIRoot in mro


def test_mi_argument_upcasts_with_pointer_adjustment():
    d = t.MID()
    b_addr, c_addr, root_b_addr, root_c_addr = t.mid_base_addresses(d)

    assert t.expect_b_ref(d) == b_addr
    assert t.expect_c_ptr(d) == c_addr
    assert t.expect_root_from_b_ref(d) == root_b_addr
    assert t.expect_root_from_c_ptr(d) == root_c_addr


def test_mi_wrong_base_rejected():
    b = t.MIB()
    c = t.MIC()

    with pytest.raises(TypeError):
        t.expect_c_ptr(b)

    with pytest.raises(TypeError):
        t.expect_b_ref(c)


def test_mi_identity_stable_across_base_views():
    d = t.make_mid()
    b = t.as_b_view()
    c = t.as_c_view()

    assert id(d) == id(b)
    assert id(d) == id(c)


def test_take_a_b_c_d_matrix():
    a = t.A()
    b = t.B()
    c = t.C()
    d = t.D()

    assert t.take_a(a).name() == "A"
    with pytest.raises(TypeError):
        t.take_b(a)
    with pytest.raises(TypeError):
        t.take_c(a)
    with pytest.raises(TypeError):
        t.take_d(a)

    assert t.take_a(b).name() == "A"
    assert t.take_b(b).name() == "B"
    with pytest.raises(TypeError):
        t.take_c(b)
    with pytest.raises(TypeError):
        t.take_d(b)

    assert t.take_a(c).name() == "A"
    with pytest.raises(TypeError):
        t.take_b(c)
    assert t.take_c(c).name() == "C"
    with pytest.raises(TypeError):
        t.take_d(c)

    assert t.take_a(d).name() == "A"
    assert t.take_b(d).name() == "B"
    assert t.take_c(d).name() == "C"
    assert t.take_d(d).name() == "D"


def test_shared_ptr_roundtrip():
    d = t.D()
    assert t.take_d_shared_ptr(d).name() == "D"


def test_base_view_downcast():
    d_as_a = t.d_factory()
    dd = t.take_d(d_as_a)
    assert dd.name() == "D"


def test_dispatch_through_each_base_view():
    a = t.A()
    b = t.B()
    c = t.C()
    d = t.D()

    assert a.name() == "A"
    assert b.name() == "B"
    assert c.name() == "C"
    assert d.name() == "D"

    assert t.take_a(d).name() == "A"
    assert t.take_b(d).name() == "B"
    assert t.take_c(d).name() == "C"
    assert t.take_d(d).name() == "D"


@pytest.fixture
def mi_lifetime_clean():
    collect()
    t.mi_lifetime_reset()
    yield
    collect()
    t.mi_lifetime_reset()


def test_mi_lifetime_python_owned_destructor_once(mi_lifetime_clean):
    value = t.MITrackedLeaf()
    assert t.mi_lifetime_stats() == (1, 0, 1)

    del value
    collect()

    assert t.mi_lifetime_stats() == (1, 1, 0)


def test_mi_lifetime_shared_ownership_destructor_once(mi_lifetime_clean):
    value = t.mi_lifetime_make_shared()
    left = t.mi_lifetime_as_left(value)
    right = t.mi_lifetime_as_right(value)

    assert id(value) == id(left)
    assert id(value) == id(right)
    assert t.mi_lifetime_stats() == (1, 0, 1)

    del left, right, value
    collect()

    assert t.mi_lifetime_stats() == (1, 1, 0)


def test_mi_lifetime_regression_multi_base_views_release_once(mi_lifetime_clean):
    left = t.mi_lifetime_singleton_left()
    right = t.mi_lifetime_singleton_right()

    assert id(left) == id(right)
    assert t.mi_lifetime_stats() == (1, 0, 1)

    t.mi_lifetime_release_singleton()
    del left, right
    collect()

    assert t.mi_lifetime_stats() == (1, 1, 0)


@pytest.mark.xfail(reason="Virtual inheritance in MI shim is out-of-scope in v1")
def test_virtual_inheritance_out_of_scope_v1():
    # Mirrors Boost.Python polymorphism2.py::test_pure_virtual coverage note:
    # Q : virtual P interaction is intentionally deferred in the first MI slice.
    p = t.P()
    with pytest.raises(RuntimeError):
        p.f()

File: tests/test_multiple_inheritance.cpp
#include <nanobind/nanobind.h>
#include <nanobind/multiple_inheritance.h>
#include <nanobind/stl/shared_ptr.h>
#include <nanobind/trampoline.h>

#include <memory>

namespace nb = nanobind;

struct MIBase {
    virtual ~MIBase() = default;
    int value = 123;
};

struct MIDerived : MIBase { };
struct MIUnrelated { };

struct MIRoot {
    virtual ~MIRoot() = default;
};

struct MIB : MIRoot { };
struct MIC : MIRoot { };
struct MID : MIB, MIC { };

struct P {
    virtual ~P() = default;
    virtual const char *f() = 0;
    const char *g() { return "P::g()"; }
};

struct PyP : P {
    NB_TRAMPOLINE(P, 1);

    const char *f() override {
        NB_OVERRIDE_PURE(f);
    }
};

struct Q : virtual P {
    const char *f() override { return "Q::f()"; }
};

struct MIA {
    MIA() : x(0) { }
    virtual ~MIA() = default;
    const char *name() { return "A"; }
    int x;
};

struct MIBoost : MIA {
    MIBoost() : x(1) { }
    const char *name() { return "B"; }
    int x;
};

struct MICBoost : MIA {
    MICBoost() : x(2) { }
    virtual ~MICBoost() = default;
    const char *name() { return "C"; }
    int x;
};

struct MIDBoost : MIBoost, MICBoost {
    MIDBoost() : x(3) { }
    const char *name() { return "D"; }
    int x;
};

struct MITrackedRoot {
    virtual ~MITrackedRoot() = default;
};

struct MITrackedLeft : MITrackedRoot {
    virtual ~MITrackedLeft() = default;
};

struct MITrackedRight : MITrackedRoot {
    virtual ~MITrackedRight() = default;
};

struct MITrackedLeaf : MITrackedLeft, MITrackedRight {
    MITrackedLeaf() {
        constructed++;
        alive++;
    }

    ~MITrackedLeaf() override {
        destroyed++;
        alive--;
    }

    static int constructed;
    static int destroyed;
    static int alive;
};

int MITrackedLeaf::constructed = 0;
int MITrackedLeaf::destroyed = 0;
int MITrackedLeaf::alive = 0;

static std::shared_ptr<MITrackedLeaf> tracked_singleton;

MIA take_a(const MIA &a) { return a; }
MIBoost take_b(MIBoost &b) { return b; }
MICBoost take_c(MICBoost *c) { return *c; }
MIDBoost take_d(MIDBoost *const &d) { return *d; }
MIDBoost take_d_shared_ptr(std::shared_ptr<MIDBoost> d) { return *d; }
std::shared_ptr<MIA> d_factory() { return std::shared_ptr<MIBoost>(new MIDBoost); }

static void *identity_cast(void *ptr) noexcept { return ptr; }

static MID *mid_singleton() {
    static MID v;
    return &v;
}

static std::shared_ptr<MITrackedLeaf> tracked_make_shared() {
    return std::make_shared<MITrackedLeaf>();
}

NB_MODULE(test_multiple_inheritance_ext, m) {
    m.def("registry_miss_before_registration", []() {
        auto fn = nb::detail::nb_type_find_cast(&typeid(MIDerived), &typeid(MIBase));
        return fn == nullptr;
    });

    m.def("registry_register_and_lookup_hit", []() {
        nb::detail::nb_type_register_cast(&typeid(MIDerived), &typeid(MIBase), identity_cast);
        auto fn = nb::detail::nb_type_find_cast(&typeid(MIDerived), &typeid(MIBase));
        return fn == identity_cast;
    });

    m.def("registry_register_with_shim_helper", []() {
        nb::mi::register_cast<MIDerived, MIBase>();
        auto fn = nb::detail::nb_type_find_cast(&typeid(MIDerived), &typeid(MIBase));
        MIDerived value;
        return fn && fn(&value) == static_cast<MIBase *>(&value);
    });

    m.def("registry_unrelated_pair_is_miss", []() {
        auto fn = nb::detail::nb_type_find_cast(&typeid(MIDerived), &typeid(MIUnrelated));
        return fn == nullptr;
    });

    nb::mi::class_<MIBase>(m, "MIBase")
        .def(nb::init<>())
        .def("value", [](const MIBase &self) { return self.value; });

    nb::mi::class_<MIRoot>(m, "MIRoot").def(nb::init<>());
    nb::mi::class_<MIB, MIRoot>(m, "MIB").def(nb::init<>());
    nb::mi::class_<MIC, MIRoot>(m, "MIC").def(nb::init<>());
    nb::mi::class_<MID, nb::mi::bases<MIB, MIC>>(m, "MID").def(nb::init<>());

    m.def("mid_base_addresses", [](MID &d) {
        return nb::make_tuple(
            (uintptr_t) static_cast<MIB *>(&d),
            (uintptr_t) static_cast<MIC *>(&d),
            (uintptr_t) static_cast<MIRoot *>(static_cast<MIB *>(&d)),
            (uintptr_t) static_cast<MIRoot *>(static_cast<MIC *>(&d)));
    });

    m.def("expect_b_ref", [](MIB &b) { return (uintptr_t) &b; });
    m.def("expect_c_ptr", [](MIC *c) { return (uintptr_t) c; });
    m.def("expect_root_from_b_ref", [](MIB &b) {
        return (uintptr_t) static_cast<MIRoot *>(&b);
    });
    m.def("expect_root_from_c_ptr", [](MIC *c) {
        return (uintptr_t) static_cast<MIRoot *>(c);
    });

    m.def("make_mid", []() -> MID * { return mid_singleton(); }, nb::rv_policy::reference);
    m.def("as_b_view", []() -> MIB * { return static_cast<MIB *>(mid_singleton()); },
          nb::rv_policy::reference);
    m.def("as_c_view", []() -> MIC * { return static_cast<MIC *>(mid_singleton()); },
          nb::rv_policy::reference);

    // Boost.Python polymorphism2.cpp virtual inheritance fixture
    nb::mi::class_<P, PyP>(m, "P")
        .def(nb::init<>())
        .def("f", &P::f)
        .def("g", &P::g);

    nb::mi::class_<Q, nb::mi::bases<P>>(m, "Q")
        .def(nb::init<>())
        .def("f", &Q::f)
        .def("g", &P::g);

    // Boost.Python m1.cpp-style multiple inheritance fixture
    nb::mi::class_<MIA>(m, "A")
        .def(nb::init<>())
        .def("name", &MIA::name);

    nb::mi::class_<MIBoost, MIA>(m, "B")
        .def(nb::init<>())
        .def("name", &MIBoost::name);

    nb::mi::class_<MICBoost, MIA>(m, "C")
        .def(nb::init<>())
        .def("name", &MICBoost::name);

    nb::mi::class_<MIDBoost, nb::mi::bases<MIBoost, MICBoost>>(m, "D")
        .def(nb::init<>())
        .def("name", &MIDBoost::name);

    m.def("take_a", &take_a);
    m.def("take_b", &take_b);
    m.def("take_c", &take_c);
    m.def("take_d", &take_d);
    m.def("take_d_shared_ptr", &take_d_shared_ptr);
    m.def("d_factory", &d_factory);

    m.def("mi_lifetime_reset", []() {
        MITrackedLeaf::constructed = 0;
        MITrackedLeaf::destroyed = 0;
        MITrackedLeaf::alive = 0;
        tracked_singleton.reset();
    });

    m.def("mi_lifetime_stats", []() {
        return nb::make_tuple(MITrackedLeaf::constructed,
                              MITrackedLeaf::destroyed,
                              MITrackedLeaf::alive);
    });

    nb::mi::class_<MITrackedRoot>(m, "MITrackedRoot");
    nb::mi::class_<MITrackedLeft, MITrackedRoot>(m, "MITrackedLeft");
    nb::mi::class_<MITrackedRight, MITrackedRoot>(m, "MITrackedRight");
    nb::mi::class_<MITrackedLeaf, nb::mi::bases<MITrackedLeft, MITrackedRight>>(m,
                                                                                 "MITrackedLeaf")
        .def(nb::init<>());

    m.def("mi_lifetime_make_shared", &tracked_make_shared);
    m.def("mi_lifetime_as_left", [](std::shared_ptr<MITrackedLeaf> value) {
        return std::static_pointer_cast<MITrackedLeft>(value);
    });
    m.def("mi_lifetime_as_right", [](std::shared_ptr<MITrackedLeaf> value) {
        return std::static_pointer_cast<MITrackedRight>(value);
    });
    m.def("mi_lifetime_singleton_left", []() {
        if (!tracked_singleton)
            tracked_singleton = std::make_shared<MITrackedLeaf>();
        return std::static_pointer_cast<MITrackedLeft>(tracked_singleton);
    });
    m.def("mi_lifetime_singleton_right", []() {
        if (!tracked_singleton)
            tracked_singleton = std::make_shared<MITrackedLeaf>();
        return std::static_pointer_cast<MITrackedRight>(tracked_singleton);
    });
    m.def("mi_lifetime_release_singleton", []() { tracked_singleton.reset(); });

    m.def("shim_construct_and_call", []() {
        MIBase value;
        return value.value;
    });
}
